### PR TITLE
Add WiFi hotspot authorization permission system

### PR DESCRIPTION
**Pull request by mallexxx**
Task/Issue URL: https://app.asana.com/1/137249556945/project/1202406491309510/task/1210940150727931?focus=true  
Tech Design URL:  
CC: @ayoy

## Description

Adds WiFi hotspot authorization permission system that detects captive portals and shows a permission dialog to users. When a navigation fails, the system tests to determine if the user is behind a captive portal. If detected, it shows a permission dialog allowing users to authorize access to the captive portal page, which opens in a popup window.

Key features:
- Hotspot detection service that monitors network connectivity
- Permission dialog system for user authorization
- Captive portal popup management with automatic closure
- Auto-reload of affected tabs when connectivity is restored
- Debug menu for testing with local test server endpoints
- Comprehensive test coverage for various scenarios

## Testing Steps

**Prerequisites:**
- Build and run DEBUG version of DuckDuckGo browser
- Feature flag `hotspotDetection` should be enabled for internal users
- Access Debug menu via **Debug → WiFi Hotspot Detection**

**Test Scenario 1: Basic Hotspot Detection with Debug Server**
1. Open Debug menu → **WiFi Hotspot Detection → Local Tests Server (Redirect to Hotspot)**
2. Verify tests server starts automatically (menu shows "Stop Tests Server [Running]")
3. Turn OFF WiFi completely
4. Navigate to `https://duckduckgo.com/?q=test+search`
5. **Expected:** Page fails to load AND permission dialog appears asking "Open WiFi authentication page"
6. Click **"Allow"** in permission dialog
7. **Expected:** Popup window opens showing redirect page with localhost:8085 URL
8. Turn ON WiFi and reconnect to network
9. In Debug menu, switch to **Local Tests Server (Success)**
10. **Expected:** Popup window closes automatically and original tab reloads with search results

**Test Scenario 2: Multiple Tabs Auto-Reload**
1. Debug menu → **Local Tests Server (Redirect to Hotspot)**, turn OFF WiFi
2. Open Tab 1: `https://duckduckgo.com/?q=tab1+search` → permission dialog appears, click "Allow", popup opens
3. Open Tab 2 and Tab 3 with different searches (tabs show error pages)
4. Close Tab 1 by clicking X button on tab (popup remains open)
5. Turn ON WiFi, then Debug menu → **Local Tests Server (Success)**
6. **Expected:** Popup closes AND both remaining tabs reload automatically

**Test Scenario 3: Permission Denial**
1. Debug menu → **Local Tests Server (Hotspot)**, turn OFF WiFi
2. Navigate to any site, permission dialog appears
3. Click **"Don't Allow"**
4. **Expected:** No popup opens, tab remains on error page
5. Reload page (Cmd+R) → permission dialog appears again (not permanently blocked)

**Test Scenario 4: Normal Browsing Unaffected**
1. Debug menu → **Local Tests Server (Hotspot)**, ensure WiFi connected
2. Navigate to `https://duckduckgo.com/?q=normal+search`, `https://example.com`, `https://github.com`
3. **Expected:** All pages load normally with NO permission dialogs

**Test Scenario 5: Server vs Network Errors**
1. Debug menu → **Local Tests Server (Failure)**, WiFi connected
2. Navigate to any site → **Expected:** Server error response with NO permission dialog
3. Debug menu → **Local Tests Server (Hotspot)**, turn OFF WiFi
4. Navigate to any site → **Expected:** Network failure AND permission dialog appears

**Test Scenario 6: HTTP Errors Don't Trigger Detection**
1. Debug menu → **Firefox** (normal endpoint), WiFi connected
2. Navigate to `https://duckduckgo.com/nonexistent-page-404`, `https://httpstat.us/500`
3. **Expected:** HTTP errors (404, 500) with NO permission dialogs

**Alternative Real Network Testing:**
1. Connect to actual WiFi network with captive portal (hotel, airport, cafe)
2. Debug menu → **Firefox** (real endpoint)
3. Navigate to `https://duckduckgo.com`
4. **Expected:** Permission dialog appears, click "Allow", captive portal opens in popup
5. Complete authentication → popup closes and original tab reloads

## Impact and Risks

**Impact Level: Medium** - New feature that affects user workflows when connecting to networks with captive portals.

#### What could go wrong?

* **False positives**: Network issues could trigger hotspot detection when not on captive portal
* **ATS bypass**: HTTP exception for detectportal.firefox.com could be misused
* **Permission spam**: Multiple failed navigations could show repeated permission dialogs
* **Tab proliferation**: Users might accidentally create many captive portal tabs

**Mitigations:**

* One-shot permission (not persisted) prevents repeated dialogs
* Specific domain exception limits ATS bypass scope
* Check prevents multiple concurrent hotspot detections
* Popup management handles deduplication and automatic closure
* Auto-reload only affects tabs that detected hotspot

## Quality Considerations

**Privacy & Security:**

* Permission is one-shot only (not stored persistently)
* ATS exception limited to specific detectportal.firefox.com domain
* No user data transmitted during hotspot detection
* Follows existing permission system patterns
* Popup uses disposable Fire session for privacy

**Performance:**

* Hotspot detection only triggered on navigation failures
* HTTP request to detectportal.firefox.com is lightweight
* Background task doesn't block UI
* Prevents multiple concurrent checks
* Auto-reload only affects tabs that need it

**Architecture:**

* Follows MVVM pattern with TabExtension architecture
* Uses dependency injection for testability
* Integrates with existing PermissionModel system
* Proper singleton removal and dependency management
* Comprehensive test coverage including edge cases

**Edge Cases:**

* Network timeouts handled gracefully
* Multiple navigation failures don't trigger multiple checks
* Works with existing permission authorization flow
* Handles missing dependencies gracefully
* Multiple tabs scenario properly handled

## Notes to Reviewer

* This is a new permission type that doesn't persist (one-shot only)
* ATS exception is required for HTTP captive portal detection
* TabExtension pattern follows existing architecture
* Debug menu provides comprehensive testing capabilities
* All affected tabs auto-reload when connectivity is restored
* Singleton anti-pattern has been eliminated with proper dependency injection

### CODE REVIEW

File: macOS/DuckDuckGo/NavigationBar/View/AddressBarTextField.swift [hunk 1]
```diff
@@ -33,7 +33,7 @@ protocol AddressBarTextFieldFocusDelegate: AnyObject {
 
 final class AddressBarTextField: NSTextField {
 
-    var tabCollectionViewModel: TabCollectionViewModel! {
+    weak var tabCollectionViewModel: TabCollectionViewModel? {
```
Comment (mallexxx) [inline macOS/DuckDuckGo/NavigationBar/View/AddressBarTextField.swift:36]: this was causing some weird leak locally on Tahoe

File: macOS/DuckDuckGo/Connectivity/HotspotDetectionService.swift [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    private let checkInterval: TimeInterval = 5.0 // Check every 5 seconds
+
+    private var testURL: URL {
+#if DEBUG && !APPSTORE
+        return HotspotDetectionDebugSettings.shared.connectivityCheckURL
+#else
+        return URL(string: "http://detectportal.firefox.com/success.txt")!
+#endif
+    }
```
Comment (samsymons) [inline macOS/DuckDuckGo/Connectivity/HotspotDetectionService.swift:65]: Won't this mean that the app will always use the final URL in non-debug builds, so testing a review build etc. will be tricky?

**Review by samsymons — APPROVED** [review]
All of the testing steps worked great, nice work! I breakpointed the flow of the hotspot detection service in each of these cases to confirm that the 5 second URL test doesn't happen during normal usage, it looked good.

Only UX issue I noticed is that the button looks horizontally compressed, which affects the corner radius etc.:

Image

File: macOS/DuckDuckGo/Connectivity/HotspotDetectionService.swift (hunk 2 above)
Comment (mallexxx) [reply macOS/DuckDuckGo/Connectivity/HotspotDetectionService.swift:65]: yes, `test-server` is only bundled in `DEBUG`

Comment (samsymons) [reply macOS/DuckDuckGo/Connectivity/HotspotDetectionService.swift:65]: Ah great, I thought we were now including it in the Sparkle build for some reason - thanks!
